### PR TITLE
Create separate module named `matchbox-flatcar` that provisions flatcar only.

### PR DESCRIPTION
**Pull request by ipochi**
This commit creates a new module named `matchbox-flatcar` which does the flatcar provisioning. Removes the code from baremetal and makes change s such that baremetal uses the new module.

Contains commits from #1374.

Please only review this commit https://github.com/kinvolk/lokomotive/pull/1387/commits/04bed78685f89922696cb98c456cc3170290e6c9

### CODE REVIEW

File: assets/terraform-modules/bare-metal/flatcar-linux/kubernetes/ssh.tf [hunk 1]
```diff
@@ -5,9 +5,7 @@ resource "null_resource" "copy-controller-secrets" {
   # Without depends_on, remote-exec could start and wait for machines before
   # matchbox groups are written, causing a deadlock.
   depends_on = [
-    matchbox_group.install,
-    matchbox_group.controller,
-    matchbox_group.worker,
+    module.controller_profile
```
Comment (invidian) [inline assets/terraform-modules/bare-metal/flatcar-linux/kubernetes/ssh.tf:8]:
```suggestion
    module.controller_profile,
```

File: assets/terraform-modules/matchbox-flatcar/profiles.tf [hunk 2]
```diff
@@ -53,44 +48,32 @@ resource "matchbox_profile" "cached-flatcar-linux-install" {
...
     # profile uses -b baseurl to install from matchbox cache
-    baseurl_flag = "-b ${var.matchbox_http_endpoint}/assets/flatcar"
+    baseurl_flag = "-b ${var.http_endpoint}/assets/flatcar"
   })
 }
 
-// Kubernetes Controller profiles
-resource "matchbox_profile" "controllers" {
-  count = length(var.controller_names)
+// Node profiles
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/profiles.tf:71]: Or maybe we can remove the comment completely, it does not bring any value IMO.
```suggestion
// Node profiles.
```

File: assets/terraform-modules/matchbox-flatcar/templates/install.yaml.tmpl [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      contents: |
+        [Unit]
+        Requires=network-online.target
+        After=network-online.target
+        [Service]
+        Type=simple
+        ExecStart=/opt/installer
+        [Install]
+        WantedBy=multi-user.target
+    # Avoid using the standard SSH port so terraform apply cannot SSH until
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/templates/install.yaml.tmpl:15]:
```suggestion
    # Avoid using the standard SSH port so Terraform apply cannot SSH until
```

File: assets/terraform-modules/matchbox-flatcar/templates/install.yaml.tmpl [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    # post-install. But admins may SSH to debug disk install problems.
+    # After install, sshd will use port 22 and users/terraform can connect.
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/templates/install.yaml.tmpl:17]: It seems this file is not detected by git as moved, so I guess we can fix this kind of stuff.
```suggestion
    # After install, sshd will use port 22 and users/Terraform can connect.
```

File: assets/terraform-modules/matchbox-flatcar/variables.tf [hunk 5]
```diff
@@ -0,0 +1,74 @@
+# bare-metal
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/variables.tf:1]: I think we can remove this one too.
```suggestion
```

File: assets/terraform-modules/matchbox-flatcar/variables.tf [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+variable "ssh_keys" {
+  type        = list(string)
+  description = "SSH public keys for user 'core'"
+}
+
+variable "ignition_clc_config" {
+  type        = string
+  description = "Ignition CLC snippets to include in the configuration."
+  #  default     = []
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/variables.tf:63]: This should be restored? Or removed :)

File: assets/terraform-modules/matchbox-flatcar/variables.tf [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+variable "node_name" {
+  type        = string
+  description = "Ordered list of node names (e.g. [node1,node2])"
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/variables.tf:68]: This is not a list.

File: assets/terraform-modules/matchbox-flatcar/variables.tf [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+variable "node_mac" {
+  type        = string
+  description = "Ordered list of node identifying MAC addresses (e.g. [52:54:00:a1:9c:ae])"
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/variables.tf:73]: Ditto list

File: assets/terraform-modules/matchbox-flatcar/variables.tf [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+variable "install_to_smallest_disk" {
+  description = "Install Flatcar Container Linux to the smallest disk."
+  type        = bool
+  default     = false
+}
+
+variable "ssh_keys" {
+  type        = list(string)
+  description = "SSH public keys for user 'core'"
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/variables.tf:57]: Would be nice to add missing dots.
```suggestion
  description = "SSH public keys for user 'core'."
```

File: assets/terraform-modules/matchbox-flatcar/variables.tf [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+variable "install_disk" {
+  type        = string
+  default     = "/dev/sda"
+  description = "Disk device to which the install profiles should install the operating system (e.g. /dev/sda)"
+}
+
+variable "container_linux_oem" {
+  type        = string
+  default     = ""
+  description = "DEPRECATED: Specify an OEM image id to use as base for the installation (e.g. ami, vmware_raw, xen) or leave blank for the default image"
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/variables.tf:40]: It seems type, default, description order is mixed between values. We could be consistent here. Given that we should always have `type` and `description`, I'd make it `type`, `description`, `default`.

File: assets/terraform-modules/matchbox-flatcar/versions.tf [hunk 11]
```diff
@@ -0,0 +1,20 @@
+# Terraform version and plugin versions
+
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/versions.tf:2]:
```suggestion
```

**Review by invidian — APPROVED** [review]
Just some nits, otherwise looks good :+1: Really nice improvement.

File: assets/terraform-modules/matchbox-flatcar/templates/install.yaml.tmpl [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+            -o "${container_linux_oem}" \
+            ${baseurl_flag} \
+            -i ignition.json
+          udevadm settle
+          systemctl reboot
+passwd:
+  users:
+    - name: core
+      ssh_authorized_keys: ${ssh_keys}
+
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/templates/install.yaml.tmpl:51]: Trailing newline?

File: assets/terraform-modules/matchbox-flatcar/variables.tf [hunk 13]
```diff
@@ -0,0 +1,72 @@
+variable "http_endpoint" {
+  type        = string
+  description = "Matchbox HTTP read-only endpoint (e.g. http://matchbox.example.com:8080)"
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/variables.tf:3]: Can we have `.` at the end of all descriptions please?

**Review by invidian — APPROVED** [review]

File: assets/terraform-modules/matchbox-flatcar/variables.tf [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+  description = "Additional kernel arguments to provide at PXE boot."
+  type        = list(string)
+  default     = []
+}
+
+variable "install_to_smallest_disk" {
+  description = "Install Flatcar Container Linux to the smallest disk."
+  type        = bool
+  default     = false
+}
```
Comment (invidian) [inline assets/terraform-modules/matchbox-flatcar/variables.tf:52]: Type and description fields are in the wrong order here still.

**Review by invidian — DISMISSED** [review]

File: assets/terraform-modules/matchbox-flatcar/variables.tf (hunk 14 above)
Comment (ipochi) [reply assets/terraform-modules/matchbox-flatcar/variables.tf:52]: oops, my bad. I wasn't thorough.

**Review by invidian — APPROVED** [review]

**Review by surajssd — APPROVED** [review]
Another LGTM :laughing:

**Pull request merged**